File: sdk/npm/src/providers/uncorreotemporal.ts
import * as https from 'https';
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';

const CHANNEL: Channel = 'uncorreotemporal';
const HOST = 'uncorreotemporal.com';
const ORIGIN = `https://${HOST}`;
const IPV4_AGENT = new https.Agent({ family: 4 });

interface CreateResponse {
  address?: string;
  expires_at?: string;
  session_token?: string;
}

interface ListMessage {
  id?: string;
  from_address?: string;
  to_address?: string;
  subject?: string;
  received_at?: string;
  is_read?: boolean;
  has_attachments?: boolean;
}

interface DetailMessage extends ListMessage {
  body_text?: string;
  body_html?: string;
  attachments?: unknown[];
}

const HEADERS: Record<string, string> = {
  Accept: 'application/json',
  Origin: ORIGIN,
  Referer: `${ORIGIN}/`,
  'User-Agent': 'Mozilla/5.0',
};

function requestJSON<T>(method: 'GET' | 'POST', path: string, token?: string): Promise<T> {
  const headers: Record<string, string> = { ...HEADERS };
  if (method === 'POST') headers['Content-Type'] = 'application/json';
  if (token) headers['X-Session-Token'] = token;

  return new Promise<T>((resolve, reject) => {
    const req = https.request(
      {
        protocol: 'https:',
        hostname: HOST,
        path,
        method,
        headers,
        agent: IPV4_AGENT,
        timeout: 15000,
      },
      (res) => {
        const chunks: Buffer[] = [];
        res.on('data', (chunk: Buffer) => chunks.push(chunk));
        res.on('end', () => {
          const body = Buffer.concat(chunks).toString('utf8');
          const status = res.statusCode || 0;
          if (status < 200 || status >= 300) {
            reject(new Error(`uncorreotemporal http ${status}`));
            return;
          }
          try {
            resolve(JSON.parse(body) as T);
          } catch (err) {
            reject(err);
          }
        });
      },
    );
    req.on('timeout', () => req.destroy(new Error('uncorreotemporal timeout')));
    req.on('error', reject);
    req.end();
  });
}

function flatten(raw: DetailMessage | ListMessage, recipient: string): Record<string, unknown> {
  return {
    id: raw.id || '',
    from: raw.from_address || '',
    to: raw.to_address || recipient,
    subject: raw.subject || '',
    text: (raw as DetailMessage).body_text || '',
    html: (raw as DetailMessage).body_html || '',
    date: raw.received_at || '',
    isRead: Boolean(raw.is_read),
    attachments: Array.isArray((raw as DetailMessage).attachments) ? (raw as DetailMessage).attachments : [],
  };
}

export async function generateEmail(): Promise<InternalEmailInfo> {
  const data = await requestJSON<CreateResponse>('POST', '/api/v1/mailboxes');
  const email = String(data.address || '').trim();
  const token = String(data.session_token || '').trim();
  if (!email || !email.includes('@') || !token) {
    throw new Error('uncorreotemporal: invalid mailbox response');
  }
  return {
    channel: CHANNEL,
    email,
    token,
    expiresAt: data.expires_at,
  };
}

async function fetchDetail(email: string, token: string, messageID: string): Promise<DetailMessage | null> {
  try {
    return await requestJSON<DetailMessage>(
      'GET',
      `/api/v1/mailboxes/${encodeURIComponent(email)}/messages/${encodeURIComponent(messageID)}`,
      token,
    );
  } catch {
    return null;
  }
}

export async function getEmails(token: string, email: string): Promise<Email[]> {
  const sessionToken = String(token || '').trim();
  const address = String(email || '').trim();
  if (!sessionToken) throw new Error('uncorreotemporal: empty session token');
  if (!address) throw new Error('uncorreotemporal: empty email');

  const rows = await requestJSON<ListMessage[]>(
    'GET',
    `/api/v1/mailboxes/${encodeURIComponent(address)}/messages?limit=50`,
    sessionToken,
  );
  const emails: Email[] = [];
  for (const row of Array.isArray(rows) ? rows : []) {
    const id = String(row.id || '').trim();
    const detail = id ? await fetchDetail(address, sessionToken, id) : null;
    emails.push(normalizeEmail(flatten(detail || row, address), address));
  }
  return emails;
}

File: sdk/go/provider/uncorreotemporal.go
package provider

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net"
	stdhttp "net/http"
	"net/url"
	"strings"
	"time"
)

const uncorreotemporalBase = "https://uncorreotemporal.com/api/v1"

type uncorreotemporalCreateResponse struct {
	Address      string `json:"address"`
	ExpiresAt    string `json:"expires_at"`
	SessionToken string `json:"session_token"`
}

var uncorreotemporalHTTPClient = &stdhttp.Client{
	Timeout: 15 * time.Second,
	Transport: &stdhttp.Transport{
		Proxy: stdhttp.ProxyFromEnvironment,
		DialContext: func(ctx context.Context, network, address string) (net.Conn, error) {
			dialer := &net.Dialer{Timeout: 15 * time.Second}
			return dialer.DialContext(ctx, "tcp4", address)
		},
	},
}

func uncorreotemporalJSON(method, u, token string, out any) error {
	var body io.Reader
	if method == stdhttp.MethodPost {
		body = strings.NewReader("")
	}
	req, err := stdhttp.NewRequest(method, u, body)
	if err != nil {
		return err
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Origin", "https://uncorreotemporal.com")
	req.Header.Set("Referer", "https://uncorreotemporal.com/")
	req.Header.Set("User-Agent", "Mozilla/5.0")
	if method == stdhttp.MethodPost {
		req.Header.Set("Content-Type", "application/json")
	}
	if token != "" {
		req.Header.Set("X-Session-Token", token)
	}
	resp, err := uncorreotemporalHTTPClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("uncorreotemporal http %d", resp.StatusCode)
	}
	return json.Unmarshal(data, out)
}

func uncorreotemporalFlatten(raw map[string]any, recipient string) map[string]any {
	return map[string]any{
		"id":          raw["id"],
		"from":        raw["from_address"],
		"to":          uncorreotemporalFirstNonEmpty(raw["to_address"], recipient),
		"subject":     raw["subject"],
		"text":        raw["body_text"],
		"html":        raw["body_html"],
		"date":        raw["received_at"],
		"isRead":      raw["is_read"],
		"attachments": raw["attachments"],
	}
}

func uncorreotemporalFirstNonEmpty(value any, fallback string) any {
	if s, ok := value.(string); ok && strings.TrimSpace(s) != "" {
		return s
	}
	return fallback
}

func UncorreotemporalGenerate() (*CreatedMailbox, error) {
	var data uncorreotemporalCreateResponse
	if err := uncorreotemporalJSON(stdhttp.MethodPost, uncorreotemporalBase+"/mailboxes", "", &data); err != nil {
		return nil, err
	}
	email := strings.TrimSpace(data.Address)
	token := strings.TrimSpace(data.SessionToken)
	if email == "" || !strings.Contains(email, "@") || token == "" {
		return nil, fmt.Errorf("uncorreotemporal: invalid mailbox response")
	}
	return &CreatedMailbox{
		Channel:   "uncorreotemporal",
		Email:     email,
		Token:     token,
		ExpiresAt: data.ExpiresAt,
	}, nil
}

func uncorreotemporalFetchDetail(email, token, messageID string) (map[string]any, error) {
	var raw map[string]any
	u := fmt.Sprintf("%s/mailboxes/%s/messages/%s", uncorreotemporalBase, url.PathEscape(email), url.PathEscape(messageID))
	if err := uncorreotemporalJSON(stdhttp.MethodGet, u, token, &raw); err != nil {
		return nil, err
	}
	return raw, nil
}

func UncorreotemporalGetEmails(token, email string) ([]NormEmail, error) {
	sessionToken := strings.TrimSpace(token)
	address := strings.TrimSpace(email)
	if sessionToken == "" {
		return nil, fmt.Errorf("uncorreotemporal: empty session token")
	}
	if address == "" {
		return nil, fmt.Errorf("uncorreotemporal: empty email")
	}
	u := fmt.Sprintf("%s/mailboxes/%s/messages?limit=50", uncorreotemporalBase, url.PathEscape(address))
	var rows []map[string]any
	if err := uncorreotemporalJSON(stdhttp.MethodGet, u, sessionToken, &rows); err != nil {
		return nil, err
	}
	out := make([]NormEmail, 0, len(rows))
	for _, row := range rows {
		raw := row
		if id := strings.TrimSpace(fmt.Sprint(row["id"])); id != "" {
			if detail, err := uncorreotemporalFetchDetail(address, sessionToken, id); err == nil {
				raw = detail
			}
		}
		out = append(out, NormalizeMap(uncorreotemporalFlatten(raw, address), address))
	}
	return out, nil
}
